Adding runlogs, report mode and generic reports

cr opened the pull request:
This is highly WIP. PR is just for making the ongoing work more transparent and to invite discussion and feedback.

cr commented:
Serious progress. Functionality is almost complete. A few highlights:
* All the runs now **exclusively** log into the run log database. The `--reportdir` argument is gone.
* There's a new `log` subcommand that deals with the run log database, for listing and deleting runs, and for generating reports etc.
* Logs are selected by a -i/--include first, then -e/--exclude logic that can handle certain keywords as well as individual log handles.
* Get a JSON dump of the result of your last run with `tlscanary log -i 1 -a json`
* The legacy HTML report is accessible via the `htmlreport` log action. It now works incrementally and will only add those runs to the report output directory that have not already been included.

What's still missing:
* Certificate chains are blindly written to the log instead of being extracted to the CertDB first. This is not a functional limitation, but logs with many results may get very large very quickly.
* The `jsonreport` log action is not implemented, yet. It is supposed to write a complete JSON-based log directory, much like the `htmlrepot` log action, with certificates broken out into files etc.
* `htmlreport` only considers logs by the `regression` mode. Are other modes really required? They seem like a half-baked afterthought, and their results are now accessible via the `json` log action.

mwobensmith commented:
Initial feedback, more coming.

* In log output, why is `download_url` for Fx builds `null`?
* What happened to the short error message that we used to have for bad URLs? We now have the "raw error"string, but this is harder to use for display purposes.
* When I invoke the log command, I see this:
    * `2017-07-06 18:14:04 WARNING MainThread tlscanary.modes.log Skipping incomplete log <tlscanary.runlog.RunLog object at 0x103334b50>`
    * Can we remove the toString for the `RunLog` object, since it doesn't display useful information?
* The current canary updates runs.txt, but we are moving away from this type of workflow. It appears that the default log output for stdout has similar functionality to the current runs.txt, in that it lists past runs. We should explore either getting rid of runs.txt - and its index page - or somehow using the log output to autogenerate one.

cr commented:
> In log output, why is download_url for Fx builds null?

I already fixed that locally. The code was simply missing. The new format uses a `package_origin` key which can hold the URL or a file or directory.

> What happened to the short error message that we used to have for bad URLs? We now have the "raw error"string, but this is harder to use for display purposes.

You're lookig at the raw, unprocessd state data. The HTML logging adds a huge deal of post-processing.



> 17-07-06 18:14:04 WARNING MainThread tlscanary.modes.log Skipping incomplete log <tlscanary.runlog.RunLog object at 0x103334b50>

You obviously have incomplete logs around. If you want to get rid of the warning, run `tlscanary log -i incomplete -a delete --really`

> Can we remove the toString for the RunLog object, since it doesn't display useful information?

To be honest, I don't know what you are referring to. Please restate your proposal.



> The current canary updates runs.txt, but we are moving away from this type of workflow. It appears that the default log output for stdout has similar functionality to the current runs.txt, in that it lists past runs. We should explore either getting rid of runs.txt - and its index page - or somehow using the log output to autogenerate one.

Not sure if I follow. The runs.txt is updated for backward compatibility. It can't be autogenerated from scratch because your local run log might not contain all the runs that were ever written to runs.txt. That's why it only ever adds to existing HTML reports. In any case this sound like it's something for a later increment.

cr commented:
I pushed a few more commits, making this now RTL from my side. What changed:

* Slight tweaks to the log format
  * Removed camel case where data was not generated by JS, but in the Python world.
  * Added full dump of `Ci.nsINSSVersion` and `application.ini::App` to candidate metadata
  * Bumped log format revision from 1 to 2
* There's now a log compatibility check, where compatible is when the format revision matches. Incompatible logs can be dumped with the `json` log action, but no other processing is attempted. Use `tlscanary log -i incompatible -a delete --really` to get rid of them.
* Fixed a regression where srcupdates would not be saved to disk.
* Fixed the runlog tests.

mwobensmith commented:
Regarding this:
* `2017-07-06 18:14:04 WARNING MainThread tlscanary.modes.log Skipping incomplete log <tlscanary.runlog.RunLog object at 0x103334b50>`

What I was trying to say is that the part between open/close brackets - which is the representation of the incomplete log - is kind of unnecessary. Or at least not very helpful, as this is the default toString of a Python object. Would it be better here to display the log's name instead?

mwobensmith commented:
Regarding my questions about runs.txt and our new log output, it is indeed something for a later date. 

Since we are now writing pure JSON logs to represent individual runs, we probably want to move to pure JSON for this file as well. I'll file another issue for this.

cr commented:
Right. m) Fixed by using `log_name` instead of `log` for generating the warning.

mwobensmith commented:
Looks great! Let's land this.

cr commented:
Okay. Rebased, squashed, and merging...
